### DISCUSSION TITLE
Implement new depthmap called Marigold

### DISCUSSION THREAD

**Discussion by loawizard** [nagadomi/nunif#96] [Ideas]
Hello 
I wonder if it could be possible to have the depthmaps from Marigold implemented in Iw3?

here is a link to GitHub and demo

https://github.com/prs-eth/marigold
https://huggingface.co/spaces/prs-eth/marigold-lcm

Thank you so much!

**Comment by donrikk** [comment]
https://huggingface.co/spaces/prs-eth/marigold-lcm 

Update in the newest marigold depth map model seems it may be even better than depth anything I’ll have some sample pictures up in a few but here is the huggingface link to test out images and it even seems to be able to do videos aswell which is why im sharing this with you to see if u could impliment it! This is there latest fastest model the creator just got back to me and shared this with me he also answers my question. He says it is possible to code this to make it do image batching so technically this can be implemented into pictures and videos he just hasn’t done that with it but he says it’s possible! here are some examples of how accurate marigold is! lcm model is the fast model btw which is the one i used for these examples.
![avatar example 2 mari (1)](https://github.com/nagadomi/nunif/assets/106938990/0ab9dbe6-e52b-40b5-8619-b53fce677bae)
![avatar example 1 mari (1)](https://github.com/nagadomi/nunif/assets/106938990/0c005cd7-10cd-48a7-a44a-8ba3ae77ab0c)
![avatar example 3 mari (1)](https://github.com/nagadomi/nunif/assets/106938990/90dba7c5-1059-469c-89a7-b81a7bd973cf)

**Reply by donrikk** [reply]
and all of those examples are at 3 depth in 3Dcombinewith a convergence of .1 or in ur program it would be the equicalent to 1.0 convergence

**Comment by donrikk** [comment]
here is also the examples of the depth maps themselves! as u can see these maps are crazy accurate!
![av test3_1 3 1 (1)](https://github.com/nagadomi/nunif/assets/106938990/b0e28530-2d29-497e-be8e-7235379851b4)
![av test2_1 2 1 (1)](https://github.com/nagadomi/nunif/assets/106938990/1a78e12c-79af-409b-8b6d-871659e08037)
![av test4_1 4 1 (1)](https://github.com/nagadomi/nunif/assets/106938990/8dc57cb4-ccf6-4596-83b5-c01bd8830ba2)

**Comment by donrikk** [comment]
now i do think the maps could use some tuning for video, it sometimes picks up to much in the image and can make distrant object seem unnatural so something like an autogain and contrast adjustment may be needed. but if u just add it as is we would be happy to test it out for u and give feedback!

**Comment by donrikk** [comment]
this example is a bit more on the normal side for most movie scenes and this is at only 1 depth, these maps could give usthe ability to get amazing 3D videos even at 1 depth!

![meh](https://github.com/nagadomi/nunif/assets/106938990/6a7e0775-ff22-44bb-bd2f-8b0c9c3f2b37)

**Comment by nagadomi** [comment]
I know of it, but have not implemented it because I did not think it is a practical model.
LCM version is probably improving on them, so it may be worth reviewing, but I'm not too enthusiastic about it.

**Reply by donrikk** [reply]
i share the same sentiment but it would be nice to test as it seems they are tryin to add video demo capabilties. honestly i think combining marigold and depth anything together like the zoedepth and anthing model are could hold some interesting benefits but i wouldnt wven know how t oachieve that but it must be possible as its done with zoe depth and anything model. especially this faster model it doesnt do to much for the foreground but it really helps with background details and edges i wish thre was a way to get the benefit of the background enhancement of marigold with the foreground accuracy of depth anything.

**Reply by nagadomi** [reply]
The original marigold is over 100x slower than DepthAnything.
LCM version is 10~20 times faster than that. But it is still very slow.

Maybe optimizing the batch size and resolution will make it much faster.
Once their video demo is released, we can get a more detailed processing time.

**Reply by nagadomi** [reply]
I have tried the video demo locally.
It was 0.9 FPS for 720x720 video (only depth, Linux RTX3070ti). With iw3 + Any_B, it is 40 FPS (depth + stereo generation).
It is slow, but faster than I thought.
Also, background depth is quite flickering.
You can try it by cloning the above huggingface space repository locally.

**Comment by gituser123456789000** [comment]
The Marigold depth maps are not accurate by the way, the mid-range is too bright, which causes a 'bulge', or 'hill' or 'wave' effect in the mid-range of the 3D effect.. this makes Marigold's maps very unnatural, which is easily seen if you compare Depth-Anything's results next to it side by side, back and forth

The only good thing about Marigold currently is the depth it pulls out of the background of an image. Through testing we've combined the foreground and the mid-range of Depth-Anything with the background of Marigold. The process hasn't been perfected, and is likely different for every image, but it can give excellent results. Depth-Anything's backgrounds are flat in comparison

**Reply by gituser123456789000** [reply]
Anaglyph examples, for quick, easy viewing

Depth-Anything:
https://media.discordapp.net/attachments/1218511652241150099/1222024375616802967/3dcombineDEPTHANYTHINGanaglyph.jpg

Marigold:
https://media.discordapp.net/attachments/1218511652241150099/1222024375935697006/3dcombineMARIGOLDanaglyph.jpg

Custom, combined Depth-Anything foreground and mid-range with Marigold's background details:
https://media.discordapp.net/attachments/1218511652241150099/1222055416503209995/ggAnaglyph.jpg

Not meant to be perfection, just a proof of concept make by an amateur.

You can see Depth-Anything is very accurate and easy on the eyes, but the background looks flat.

Marigold's mid-range and foreground are distorted due to being too bright. See the bottoms of trees in the midrange are stretched forward.. the tree directly in the center is an example, also the octopus tentacle looking plant on the left side. The Stones or crystals in the middle-range to foreground are distorted, with the tops of the stones being deeper into the image compared to the bottoms. Everything is distorted.

The one good thing about Marigold is the 3D that's brought out of the background.. the birds for example.. the floating isalnds in the sky.

So the third image combines the strength of both models, providing a natural 3d effect from front to back.

**Reply by gituser123456789000** [reply]
This is the process:

I'm not sure how we can automate this, but I've tested with other images and results are along the same lines.. Marigold pulls out more background detail while performing poorly in the midrange, due to being too bright for its natural proper depth.

First I save the black and white Marigold results... then you have to invert the colors, because Marigold outputs in black front / white rear instead fo the opposite for pretty much every other depth map program.

In paint.net, I use the Magic Wand to select a bright white area.. change the Flood Mode on top to Global (to select all of the whitest, brightest parts of the Marigold depth map)... then adjust the Tolerance % on top to leave only the background details, like pictured here as an example:
https://media.discordapp.net/attachments/1218511652241150099/1222040713479327805/paintdotnet.PNG

Then delete all of that foreground and mid-range with the delete key on your keyboard

Then you make a layer over the Depth-Anything map (or you could have done this step first).. and copy/paste what's left of the Marigold background on top of it:
https://media.discordapp.net/attachments/1218511652241150099/1222053355300585584/og.png

Now if you know depth maps, you can see the background is too bright and would become foreground.. so we need to fix the brightness between the two depth maps. Darkening the background is a possibility, or part of the possibility.. it will depend on the image, but as you darken, you'll start losing the detail you want.. so in this case, we go to the Depth-Anything layer and brighten it until everything looks to be at the correct depth:
https://media.discordapp.net/attachments/1218511652241150099/1222053355866554378/og2.png

Then I merged the layers and darkened the combined depth maps image, to make the background more black (and therefore depth for the 3d effect). This will vary by image:
https://media.discordapp.net/attachments/1218511652241150099/1222054131146031104/og3.png

And this gave the result of Depth-Anything's high accuracy from the front to near the background, along with Marigold's detailed background depth. Here are the results, combined with the original image, in SBS, crosseyed, and anaglyph.

Original image used:
https://images.squarespace-cdn.com/content/v1/5a7f41ad8fd4d236a4ad76d0/1670121236520-E89JZMD2FV78P3TB6WO0/DAK_Pandora_LowAngleNight_DC_v04MerchHD.jpg

SBS:
https://media.discordapp.net/attachments/1218511652241150099/1222055415483994182/ggSBS.jpg

Crosseyed:
https://media.discordapp.net/attachments/1218511652241150099/1222055416133845032/ggCross.jpg

Anaglyph:
https://media.discordapp.net/attachments/1218511652241150099/1222055416503209995/ggAnaglyph.jpg

If someone can figure out how to automate this process.. or we get a new depth map program that can do this (be accurate throughout, while also pulling out background depth detail).. we're going to have a nice upgrade to our 2D to 3D conversions.

**Reply by nagadomi** [reply]
I can't see some of the images, so this may contain some misguided comments.

DepthAnything outputs Reletive Depth. In fact Relative Depth is Disparity, not Depth.
Disparity is the distance between two corresponding points in the left and right images of a stereo pair.
So it is normal for it to be flatter(darker) the farther away.

If you want to enhance the background disparity, you can edit using Tone Curve.

original output of Any_B resolution=770
![depth_anything_b_770](https://github.com/nagadomi/nunif/assets/287255/b5e6b70c-3103-4e49-8599-c172e1887a00)

Editing with GIMP's Color Curves tool.
![tone_curve](https://github.com/nagadomi/nunif/assets/287255/fcb40f99-9b56-4766-a6ec-3232e49136bd)
In this example, banding artifact is occurring, but this is due to saving the image as 8-bit. iw3 handles the image as a floating point number, so this problem does not occur.

This Tone Curve editing feature is implemented in iw3 as `--mapper` option, which is used internally by `--foreground-scale` option. Currently, only curves that enhance foreground are supported.
It is easy to add negative options to `--forground-scale` that enhances the background.

**Reply by gituser123456789000** [reply]
I noticed the image links broke soon after I posted, so they should be attached below

I never got good results from using Foreground Scale. I always see unnatural distortions in the results compared to not using this setting.

**Reply by gituser123456789000** [reply]
I have an update request to add the ability to export the depth maps only. So we can see exactly what our changes are doing in the depth map, or for users who want the depth map so that they can modify them manually.

Edited To Add: Now that I've played with the Curve option.. I see this would probably be very difficult to get right just by putting in a value into the GUI. You really need to see the real-time effect that the changing Curve values make to the depth map. Even being able to see the curve making changes in real-time, it is still difficult to pull out extra detail without harming other parts of the depth map. So my second opinion on a Background Scale option would be that it may be interesting to tinker with, but I think getting actual improvements from it, without distortions, probably isn't likely to happen.

I see you had good looking results with this method though, which can't be argued with, so it seems you can probably pull this off.

**Reply by gituser123456789000** [reply]
Feedback on the dubios2 anaglyph addition: I get poor results with all images. I'm seeing dark, distorted bands in the cyan, right eye.. usually around the edges of the subject. These problem areas look to be a very bright red color.

**Comment by gituser123456789000** [comment]
![3dcombineDEPTHANYTHINGanaglyph](https://github.com/nagadomi/nunif/assets/156575428/6ec7c71d-3836-4307-966e-f04b36be0eb2)
![z3dcombineMARIGOLDanaglyph](https://github.com/nagadomi/nunif/assets/156575428/d7bb7758-f655-4f1a-9aa6-321e2d3b4a46)
![ggAnaglyph](https://github.com/nagadomi/nunif/assets/156575428/9c01c698-c9d5-437d-89d5-b46ccea6162a)
<img width="1921" alt="paintdotnet" src="https://github.com/nagadomi/nunif/assets/156575428/15be19ee-2a49-401a-b988-db0ffdb2e51c">
<img width="1917" alt="og" src="https://github.com/nagadomi/nunif/assets/156575428/b24ff5af-f035-4b47-8109-63dc2ffe1e43">
<img width="1917" alt="og2" src="https://github.com/nagadomi/nunif/assets/156575428/1cca0a07-8113-42a2-aa92-525c5a490c61">
<img width="1914" alt="og3" src="https://github.com/nagadomi/nunif/assets/156575428/849169f4-1d08-4ac0-b8e1-e937c4ffd0ea">
![ggSBS](https://github.com/nagadomi/nunif/assets/156575428/1f3486f7-7492-4c1d-8f15-ec6d52a8f8c1)
![ggCross](https://github.com/nagadomi/nunif/assets/156575428/7fd523f2-29ba-4f74-93ed-743d1bc82798)
![ggAnaglyph](https://github.com/nagadomi/nunif/assets/156575428/2e09484a-430f-483f-b30d-b4fe01596eb6)
